src/applied-upgrades-utils.ts

```diff
@@ -0,0 +1,18 @@
+export function addAppliedUpgrades(appliedUpgrades: string[] | null, upgradeName: string) {
+    if (appliedUpgrades) {
+        if (appliedUpgrades.includes(upgradeName)) {
+            return appliedUpgrades
+        }
+
+        return [...appliedUpgrades, upgradeName]
+    }
+
+    return [upgradeName]
```

Comment:
Could this be slightly clearer if you use a default value for appliedUpgrades?

```ts
export function addAppliedUpgrades(appliedUpgrades: string[] = [], upgradeName: string) {
        if (appliedUpgrades.includes(upgradeName)) {
            return appliedUpgrades
        }

        return [...appliedUpgrades, upgradeName]
}
```

Reply:
Looking at the consuming code, it handles the null case already. Simplified